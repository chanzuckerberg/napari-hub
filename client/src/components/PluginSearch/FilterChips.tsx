import { Button, Chip } from '@material-ui/core';
import { Close } from '@material-ui/icons';
import { isEmpty } from 'lodash';

import { useSearchState } from '@/context/search';
import { FilterFormState } from '@/context/search/filter.types';

/**
 * Labels to render for a particular state key.
 */
const KEY_LABELS: Record<keyof FilterFormState, string> = {
  developmentStatus: 'Development status',
  license: 'License',
  operatingSystems: 'Operating system',
  pythonVersions: 'Python version',
};

/**
 * Labels to render for a particular sub-state key.
 *
 * TODO Find a better way to add typing for this to prevent discrepancies
 * between the state data and the labels.
 */
const SUBKEY_LABEL: Record<string, string | undefined> = {
  // Dev Status labels
  onlyStablePlugins: 'Stable',

  // License labels
  onlyOpenSourcePlugins: 'Open Source',

  // OS labels
  linux: 'Linux',
  mac: 'macOS',
  windows: 'Windows',
};

interface Props {
  className?: string;
}

/**
 * Component that renders a list of chips for each enabled filter. Each chip
 * includes a button to disable the filter. It's also possible to clear all the
 * filters by clicking the clear all filters button.
 */
export function FilterChips({ className }: Props) {
  const { filter } = useSearchState() ?? {};

  if (isEmpty(filter?.chips)) {
    return null;
  }

  return (
    <div className={className}>
      {/* Clear filters button */}
      <Button
        classes={{ label: 'underline' }}
        className="inline-block mr-2"
        onClick={() => filter?.clearAll()}
      >
        Clear all filters
      </Button>

      {/* Chip list */}
      <ul className="inline flex-wrap gap-2">
        {filter?.chips.map(({ id, key, subKey }) => (
          <li className="inline-block my-1 ml-1" key={id}>
            <Chip
              label={
                <>
                  <span>{KEY_LABELS[key]}</span>
                  <span className="font-bold ml-1">
                    {SUBKEY_LABEL[subKey] ?? subKey}
                  </span>
                </>
              }
              deleteIcon={<Close className="text-black fill-current w-4 h-4" />}
              onDelete={() => filter?.removeChip(key, subKey)}
            />
          </li>
        ))}
      </ul>
    </div>
  );
}
