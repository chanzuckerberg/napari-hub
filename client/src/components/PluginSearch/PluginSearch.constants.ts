/**
 * ID used for the plugin search container. This is useful for scrolling to the
 * the search container.
 */
export const PLUGIN_SEARCH_ID = 'pluginSearch';
