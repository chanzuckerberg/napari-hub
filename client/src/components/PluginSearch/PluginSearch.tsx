import { useSearchState } from '@/context/search';

import { ColumnLayout } from '../common';
import { MediaFragment } from '../common/media';
import { SearchResult } from './SearchResult';

/**
 * Component for rendering the plugin search bar and results.
 */
export function PluginSearch() {
  const { results = [] } = useSearchState() ?? {};

  return (
    <ColumnLayout>
      {/* TODO filter column */}
      <MediaFragment greaterThanOrEqual="3xl">
        <div />
      </MediaFragment>

      {/* Search results */}
      <div>
        <h2 className="font-bold text-2xl mb-6">Browse plugins</h2>
        {results.map(({ plugin }) => (
          <SearchResult key={plugin.name} plugin={plugin} />
        ))}
      </div>
    </ColumnLayout>
  );
}
