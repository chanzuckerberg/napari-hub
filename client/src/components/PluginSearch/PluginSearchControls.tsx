export function PluginSearchControls() {
  return (
    <div className="col-span-2">
      <h1>TODO search filters</h1>
    </div>
  );
}
