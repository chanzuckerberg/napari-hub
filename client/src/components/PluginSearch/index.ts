export * from './PluginSearch';
export { PLUGIN_SEARCH_ID } from './PluginSearch.constants';
