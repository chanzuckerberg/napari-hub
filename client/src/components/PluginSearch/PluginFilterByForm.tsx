import { FormLabel } from '@material-ui/core';

import { Accordion } from '@/components/common';
import { Media } from '@/components/common/media';
import { useSearchState } from '@/context/search';

import { PluginFilterBySection } from './PluginFilterBySection';

const SECTION_LABELS: Record<string, string | undefined> = {
  // Operating System
  linux: 'Linux',
  mac: 'macOS',
  windows: 'Windows',

  // Development Status
  onlyStablePlugins: 'Only show stable plugins',

  // License
  onlyOpenSourcePlugins: 'Only show plugins with open source licenses',
};

/**
 * Component for the form for selecting the plugin filter type.
 */
function FilterForm() {
  const { filter } = useSearchState() ?? {};
  const sections = [
    {
      title: 'Python versions',
      state: filter?.state.pythonVersions,
      setState: filter?.setPythonVersion,
    },
    {
      title: 'Operating system',
      state: filter?.state.operatingSystems,
      setState: filter?.setOperatingSystem,
    },
    // TODO Uncomment when we figure out what to do with the dev status filter
    // {
    //   title: 'Development status',
    //   state: filter?.state.developmentStatus,
    //   setState: filter?.setDevelopmentStatus,
    // },
    {
      title: 'License',
      state: filter?.state.license,
      setState: filter?.setLicense,
    },
  ];

  return (
    <div className="grid grid-cols-1 gap-6 screen-600:grid-cols-2 screen-875:grid-cols-1">
      {/* Only show label on larger screens. This is because the Accordion already includes a title. */}
      <Media greaterThanOrEqual="screen-875">
        <FormLabel
          className="uppercase text-black font-semibold text-sm"
          component="legend"
          focused={false}
        >
          Filter By
        </FormLabel>
      </Media>

      {sections.map((section) => (
        <PluginFilterBySection
          key={section.title}
          title={section.title}
          filters={Object.entries(section.state ?? {}).map(
            ([key, enabled]: [string, boolean]) => ({
              enabled,
              label: SECTION_LABELS[key] ?? key,
              setEnabled: (nextEnabled: boolean) =>
                section.setState?.({ [key]: nextEnabled }),
            }),
          )}
        />
      ))}
    </div>
  );
}

/**
 * Renders the plugin filter form. For smaller screen sizes (< 875px), an
 * expandable accordion layout is used. For larger screens, the filter form is
 * rendered as-is.
 */
export function PluginFilterByForm() {
  const form = <FilterForm />;

  return (
    <>
      <Media lessThan="screen-875">
        <Accordion title="Filter By">{form}</Accordion>
      </Media>

      <Media greaterThanOrEqual="screen-875">{form}</Media>
    </>
  );
}
