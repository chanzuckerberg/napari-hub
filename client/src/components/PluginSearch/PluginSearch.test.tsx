import { render, screen } from '@testing-library/react';
import { set } from 'lodash';
import { NextRouter, useRouter } from 'next/router';

import {
  PluginSearchProvider,
  SearchQueryParams,
  SearchSortType,
} from '@/context/search';
import { URLParameterStateProvider } from '@/context/urlParameters';
import pluginIndex from '@/fixtures/index.json';

import { PluginSearch } from './PluginSearch';

function mockSearch(query = '') {
  type Replace = NextRouter['replace'];
  const replace = jest
    .fn<ReturnType<Replace>, Parameters<Replace>>()
    .mockReturnValue(Promise.resolve(true));

  const params = [
    [SearchQueryParams.Search, query],
    [SearchQueryParams.Sort, SearchSortType.Relevance],
  ];

  (useRouter as jest.Mock).mockReturnValue({
    asPath: `/?${new URLSearchParams(params).toString()}`,
    pathname: '/',
    query: params.reduce(
      (result, [key, value]) => set(result, key, value),
      {} as Record<string, string>,
    ),
    push: jest.fn(),
    replace,
  });

  const component = render(
    <URLParameterStateProvider>
      <PluginSearchProvider pluginIndex={pluginIndex}>
        <PluginSearch />
      </PluginSearchProvider>
    </URLParameterStateProvider>,
  );

  return { component, replace };
}

describe('<PluginSearch />', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockClear();
  });

  it('should match snapshot', () => {
    const { component } = mockSearch();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should show result with match in name', () => {
    const query = 'segment';
    mockSearch(query);

    const [title] = screen.getAllByTestId('searchResultName');
    expect(title.innerHTML).toContain(query);
  });

  it('should show result with match in summary', () => {
    const query = 'bio';
    mockSearch(query);

    const [summary] = screen.getAllByTestId('searchResultSummary');
    expect(summary.innerHTML).toContain(query);
  });

  it('should show result with match in author name', () => {
    const query = 'ziyang';
    mockSearch(query);

    const [author] = screen.getAllByTestId('searchResultAuthor');
    expect(author.innerHTML.toLowerCase()).toContain(query);
  });

  it('should support fuzzy matching', () => {
    mockSearch('neuron');

    const [author] = screen.getAllByTestId('searchResultSummary');
    expect(author.innerHTML.toLowerCase()).toContain('neuro');
  });
});
