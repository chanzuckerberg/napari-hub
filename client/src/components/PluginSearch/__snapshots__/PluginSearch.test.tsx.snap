// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PluginSearch /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex 2xl:grid 2xl:justify-center p-6 md:p-14 2xl:px-0 2xl:gap-14 2xl:grid-cols-napari-2-col 3xl:grid-cols-napari-3-col"
  >
    <div
      class="fresnel-greaterThanOrEqual-3xl"
    />
    <div>
      <h2
        class="font-bold text-2xl mb-6"
      >
        Browse plugins
      </h2>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-compressed-labels-io"
          >
            napari-compressed-labels-io
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Plugin exploring different options for reading and writing compressed and portable labels layers in napari.
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Draga Doncila 
            </span>
            (ddoncila@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/affinder"
          >
            affinder
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Quickly find the affine matrix mapping one image to another using manual correspondence points annotation
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Juan Nunez-Iglesias 
            </span>
            (juan.nunez-iglesias@monash.edu)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napping"
          >
            napping
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Control point mapping and coordination transformation using napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Jonas Windhager 
            </span>
            (jonas.windhager@uzh.ch)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-console"
          >
            napari-console
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A plugin that adds a console to napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Nicholas Sofroniew 
            </span>
            (sofroniewn@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-hdf5-labels-io"
          >
            napari-hdf5-labels-io
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Napari plugin to store set of layers in a .h5 file. Label layer are stored in a sparse representation
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Duway Nicolas Lesmes Leon 
            </span>
            (dlesmesleon@hotmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/PlatyMatch"
          >
            PlatyMatch
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          \`PlatyMatch\` allows registration of volumetric images of embryos by establishing correspondences between cells
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Manan Lalit 
            </span>
            (lalit@mpi-cbg.de)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari_video"
          >
            napari_video
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          napari plugin for reading videos.
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Jan Clemens 
            </span>
            (clemensjan@googlemail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/cellfinder-napari"
          >
            cellfinder-napari
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Efficient cell detection in large images
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Adam Tyson 
            </span>
            (code@adamltyson.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-svg"
          >
            napari-svg
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A plugin for reading and writing svg files with napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Nicholas Sofroniew 
            </span>
            (sofroniewn@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-yapic-prediction"
          >
            napari-yapic-prediction
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Napari widget plugin to perform yapic model segmentation prediction in the napari window
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Duway Nicolas Lesmes Leon 
            </span>
            (dlesmesleon@hotmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-demo"
          >
            napari-demo
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          demo plugin
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Ziyang Liu 
            </span>
            (zliu@chanzuckerberg.com)
          </li>
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Justin T. Kiggins 
            </span>
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/ome-zarr"
          >
            ome-zarr
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Implementation of images in Zarr files.
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              The Open Microscopy Team 
            </span>
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-brainreg"
          >
            napari-brainreg
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Visualise brainreg registration output
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Adam Tyson 
            </span>
            (adam.tyson@ucl.ac.uk)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-itk-io"
          >
            napari-itk-io
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          File IO with itk for napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Matt McCormick 
            </span>
            (matt.mccormick@kitware.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-imc"
          >
            napari-imc
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Imaging Mass Cytometry (IMC) file type support for napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Jonas Windhager 
            </span>
            (jonas.windhager@uzh.ch)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-mri"
          >
            napari-mri
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A simple plugin to use with napari for 3D-viewing of                  Magnetic Resonance Imaging file formats
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              SUSMITA SAHA 
            </span>
            (susmi06@yahoo.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-properties-viewer"
          >
            napari-properties-viewer
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A viewer for napari layer properties
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Kevin Yamauchi 
            </span>
            (kevin.yamauchi@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-medical-image-formats"
          >
            napari-medical-image-formats
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A Plugin in order to read medical image formats such as DICOM and NIfTI
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Marc Boucsein, Marc Buckmakowski 
            </span>
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/misic-napari-plugin"
          >
            misic-napari-plugin
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          segmentation of bacteria
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              lcb-iam-pswap-le 
            </span>
            (spanigrahi@imm.cnrs.fr)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-dv"
          >
            napari-dv
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Deltavision/MRC file reader for napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Talley Lambert 
            </span>
            (talley.lambert@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/nd2-dask"
          >
            nd2-dask
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Plugin to load nd2 data into napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Draga Doncila Pop 
            </span>
            (ddon0001@student.monash.edu)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-czifile2"
          >
            napari-czifile2
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Carl Zeiss Image (.czi) file support for napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Jonas Windhager 
            </span>
            (jonas.windhager@uzh.ch)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-dzi-zarr"
          >
            napari-dzi-zarr
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          An experimental plugin for viewing Deep Zoom Images (DZI) with napari and zarr.
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Trevor Manz 
            </span>
            (trevor.j.manz@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-animation"
          >
            napari-animation
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A plugin to make animations with napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Nicholas Sofroniew, Alister Burt, Guillaume Witz, Faris Abouakil, Talley Lambert 
            </span>
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-minimal-plugin"
          >
            napari-minimal-plugin
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A minimal plugin
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Manan Lalit 
            </span>
            (lalit@mpi-cbg.de)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-cellfinder"
          >
            napari-cellfinder
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Visualise cellfinder results with napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Adam Tyson 
            </span>
            (adam.tyson@ucl.ac.uk)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-nikon-nd2"
          >
            napari-nikon-nd2
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Opens Nikon ND2 files into napari.
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Chris Wood 
            </span>
            (cwood1967@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-em-reader"
          >
            napari-em-reader
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A napari plugin to read .em files
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Lorenzo Gaifas 
            </span>
            (brisvag@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/elastix-napari"
          >
            elastix-napari
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A toolbox for rigid and nonrigid registration of images.
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Viktor van der Valk 
            </span>
            (v.o.van_der_valk@lumc.nl)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/PartSeg"
          >
            PartSeg
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          PartSeg is python GUI for bio imaging analysis especially nucleus analysis,
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Grzegorz Bokota 
            </span>
            (g.bokota@cent.uw.edu.pl)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/brainglobe-napari-io"
          >
            brainglobe-napari-io
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Read and write files from the BrainGlobe neuroanatomy suite
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Adam Tyson 
            </span>
            (adam.tyson@ucl.ac.uk)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-aicsimageio"
          >
            napari-aicsimageio
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          AICSImageIO bindings for napari
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Jackson Maxfield Brown 
            </span>
            (jmaxfieldbrown@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-lazy-openslide"
          >
            napari-lazy-openslide
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A plugin to lazily load multiscale whole-slide images with openslide and dask.
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Trevor Manz 
            </span>
            (trevor.j.manz@gmail.com)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/napari-btrack-reader"
          >
            napari-btrack-reader
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          A plugin to load btrack files
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Alan R. Lowe 
            </span>
            (code@arlowe.co.uk)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/cellpose-napari"
          >
            cellpose-napari
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          a generalist algorithm for anatomical segmentation
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Carsen Stringer 
            </span>
            (stringerc@janelia.hhmi.org)
          </li>
        </ul>
      </article>
      <article
        class="py-5 border-t-2 border-black"
      >
        <h3
          class="inline font-bold text-lg hover:underline"
          data-testid="searchResultName"
        >
          <a
            href="/plugins/brainreg-segment"
          >
            brainreg-segment
          </a>
        </h3>
        <h4
          class="mt-2"
          data-testid="searchResultSummary"
        >
          Manual segmentation of 3D brain structures in a common anatomical space
        </h4>
        <ul
          class="mt-5"
        >
          <li
            class="my-2"
            data-testid="searchResultAuthor"
          >
            <span
              class="font-bold"
            >
              Adam Tyson, Horst Obenhaus 
            </span>
            (code@adamltyson.com)
          </li>
        </ul>
      </article>
    </div>
  </div>
</DocumentFragment>
`;
